### PR TITLE
Write custom deserializer function for Option<Cow<str>>

### PR DESCRIPTION
**Pull request by sffc**
Workaround for https://github.com/serde-rs/serde/issues/2016

@Manishearth: opinions on where this should go?

### CODE REVIEW

File: provider/core/tests/option_cow.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use std::borrow::Cow;
+use std::fmt;
+
+fn deserialize_option_cow_str<'de, D>(deserializer: D) -> Result<Option<Cow<'de, str>>, D::Error>
```
Comment (Manishearth) [inline provider/core/tests/option_cow.rs:8]: chore: Probably should have a comment saying why we have this

**Review by Manishearth — APPROVED** [review]
No opinion on where this should go. Probably worth making it a public export of provider/core? Other data structs will need this, yes?

**Comment by sffc** [comment]
I can stick it in the icu_provider module, but I'm just a little hesitant to start making icu_provider a bag of random utilities. But maybe there just isn't a better place to use it, and most call sites that need this will be depending on icu_provider.

**Comment by Manishearth** [comment]
Yeah I don't have a good answer for you, I don't have strong opinions on where it should go so use your best judgement imo.

**Comment by sffc** [comment]
2021-06-24:

- @nordzilla - I kind-of- like the idea of adding it as a separate utility.
- @Manishearth - But then it has to be published on crates.io.
- @nordzilla - Yeah, good point. So I think we should put this in icu_provider and hopefully we can delete it when the upstream bug gets fixed.
- @sffc - I filed a follow-up issue: https://github.com/unicode-org/icu4x/issues/825

**Comment by Manishearth** [comment]
I think upstreaming this is going to be unlikely and for now we should just stick this in provider_core.

**Comment by sffc** [comment]
> I think upstreaming this is going to be unlikely and for now we should just stick this in provider_core.

Acknowledged, but I made a PR for upstreaming this anyway:

https://github.com/serde-rs/serde/pull/2072

It should be noted that this only affects serde_derive, not serde itself.

**Comment by robertbastian** [comment]
I need this in #1550, so let's merge? We could put it alongside `deduplicating_array` in a `serde_utils` crate.

**Comment by sffc** [comment]
Superseded by #1556.

**Pull request closed without merging**